### PR TITLE
IP Space integration 

### PR DESCRIPTION
**Pull request by Didainius**
This PR finalizes IP Space integration to the SDK which was started in #578.

It now adds support for the integration layer which consists of:
* Additional field for **external network** to support IP spaces
* IP Space **Uplink management** for external networks with new type `IpSpaceUplink`
* IP Space **IP and Prefix Allocation** using new type `IpSpaceIpAllocation`
* IP Space **Org allocation** to define **custom quotas** using new type `IpSpaceOrgAssignment`

Additionally, a type `TaskResult` is attached to `Task` as IP Allocation Requests return values in that field.

Tested on 10.3.0 (validated test skips), 10.4.1, 10.4.2

### CODE REVIEW

File: .changes/v2.21.0/579-improvements.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+* External Network V2 now supports IP Spaces on VCD 10.4.1+ with new fields `UsingIpSpace` and
```
Comment (lvirbalas) [inline .changes/v2.21.0/579-improvements.md:1]: What about specifying exact type name you're referring with "External Network V2" too?

File: govcd/ip_space_org_assignment.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+	singleResult, err := oneOrError("Org ID", orgId, results)
+	if err != nil {
+		return nil, err
+	}
+
+	return singleResult, nil
+}
+
+// Update updates Org Assignment
```
Comment (lvirbalas) [inline govcd/ip_space_org_assignment.go:133]: This reads strange :)

File: govcd/ip_space_uplink.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+	err = client.OpenApiGetItem(apiVersion, urlRef, nil, response.IpSpaceUplink, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+// Update updates IP Space Uplink
```
Comment (lvirbalas) [inline govcd/ip_space_uplink.go:130]: Same as above, I think we should not use double "Update updates" wording.

File: govcd/ip_space_uplink.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+	err = client.OpenApiPutItem(apiVersion, urlRef, nil, ipSpaceUplinkConfig, result.IpSpaceUplink, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error updating IP Space: %s", err)
+	}
+
+	return result, nil
+}
+
+// Delete deletes IP Space Uplink
```
Comment (lvirbalas) [inline govcd/ip_space_uplink.go:158]: Hm :) "Delete deletes" - have we always used this "double-word" form?

**Review by lvirbalas — APPROVED** [review]
This looks great, only few nit picks on comments "Update updates" and "Delete deletes".

File: govcd/ip_space_org_assignment.go (hunk 2 above)
Comment (Didainius) [reply govcd/ip_space_org_assignment.go:133]: It is probably not the first time.
The point is that starting Go's function comments with the function name itself is recommended. At least some documentation tools (like `go doc`) used to automatically inject it as the first word if it wasn't it.

And then it is me being overly simple to reuse the same verb :) I will see if I can rephrase it better.

Comment (lvirbalas) [reply govcd/ip_space_org_assignment.go:133]: Maybe so?

`// Update Org Assignment`

Comment (Didainius) [reply govcd/ip_space_org_assignment.go:133]: True, that is a simpler way without putting too many words. I have changed these occurrences.

File: .changes/v2.21.0/579-improvements.md (hunk 1 above)
Comment (Didainius) [reply .changes/v2.21.0/579-improvements.md]: Fixed

File: govcd/ip_space_uplink.go (hunk 3 above)
Comment (Didainius) [reply govcd/ip_space_uplink.go]: Changed as per suggestion https://github.com/vmware/go-vcloud-director/pull/579#discussion_r1243557652

File: govcd/ip_space_uplink.go (hunk 4 above)
Comment (Didainius) [reply govcd/ip_space_uplink.go]: Changed as per suggestion https://github.com/vmware/go-vcloud-director/pull/579#discussion_r1243557652

File: govcd/openapi_endpoints.go [hunk 5]
```diff
@@ -146,6 +151,14 @@ var endpointElevatedApiVersions = map[string][]string{
 		//"35.0", // Introduced support
 		"37.0", // Added metadata support
 	},
+	types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointEdgeGateways: {
+		//"35.0", // Introduced support
+		"37.1", // Exposes computed field `UsingIpSpace` in `types.EdgeGatewayUplinks`
+	},
+	types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointIpSpaceUplinksAllocate: {
+		//"37.1", // Introduced support
+		"37.2", // Adds 'value' field
```
Comment (adezxc) [inline govcd/openapi_endpoints.go:160]: Did 37.2 introduce a new field that wasn't there before?

**Review by adezxc — APPROVED** [review]
Terrific work,  approved. Only some questions

File: govcd/openapi_endpoints.go (hunk 5 above)
Comment (Didainius) [reply govcd/openapi_endpoints.go:160]: Yep, this was not available in `37.1` (when it was introduced). This mapping will make sure that `getOpenApiHighestElevatedVersion` picks `37.2` when it is available (being on VCD 10.4.2+)

File: govcd/ip_space_allocation.go [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+	}
+
+	err = client.OpenApiDeleteItem(apiVersion, urlRef, nil, getTenantContextHeader(tenantContext))
+	if err != nil {
+		return err
+	}
+
+	if err != nil {
+		return fmt.Errorf("error deleting IP Space IP Allocation: %s", err)
+	}
```
[the remark this answers is not in the document]
Comment (Didainius) [reply govcd/ip_space_allocation.go:199]: Great spot!fixed

File: types/v56/ip_space.go [hunk 7]
```diff
@@ -200,3 +200,119 @@ type IPPrefixSequence struct {
...
+	// The type of the IP Space associated with this uplink. Possible values are: PUBLIC, PRIVATE,
+	// SHARED_SERVICES. This property is read-only.
+	IPSpaceType string `json:"ipSpaceType,omitempty"`
+	Status      string `json:"status,omitempty"`
+}
+
+// IpSpaceIpAllocationRequest is an IP Space IP Allocation request object. An IP Space IP allocation request can
+// either request a specific IP address/IP prefix or request a specific number of any free IP
+// Addresses/IP Prefixes within an IP Space. To allocate a specific IP Address or IP Prefix, the
+// value field should be used and the IP Address or Prefix should be specified. To allocate up to 5
```
Comment (dataclouder) [inline types/v56/ip_space.go:222]: Could you explain a bit this limitation? Is it because the API doesn't return more than 5?

Comment (Didainius) [reply types/v56/ip_space.go:222]: Yes. But now as you reiterate that - I am not sure if I should even note that - this is a thing that can change with versions.

File: govcd/ip_space_org_assignment.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+	"net/url"
+
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+)
+
+// IpSpaceOrgAssignment handles Custom Quotas (name in UI) for a particular Org. They complement
+// default quotas which are being set in IP Space itself.
+// The behavior of IpSpaceOrgAssignment is specific - whenever an NSX-T Edge Gateway backed by
+// Provider gateway using IP Spaces is being created - Org Assignment is created implicitly. One can
+// look that assignment by IP Space and Org to update `types.IpSpaceOrgAssignment.CustomQuotas`
```
[the remark this answers is not in the document]
Comment (Didainius) [reply govcd/ip_space_org_assignment.go:18]: Fixed.

File: govcd/task.go [hunk 9]
```diff
@@ -464,3 +464,40 @@ func buildFilterTextWithLogicalOr(filter map[string]string) string {
...
+
+// WaitForRunningTasksByName is a convenience function to query for unfinished tasks by
+// 'name' and wait until their finish
+
+// This is convenient when a particular operation needs to be found by not having any reference to
+// parent object. An exact case for it was that updating some IP Space related objects (IP Spaces,
+// IP Space Uplinks). Updating such an object sometimes results in a separate task for Route
+// Advertisement being spun up (name="ipSpaceUplinkRouteAdvertisementSync"). When such
+// task is running - other operations may fail so it is best to wait for completion of such task
+// before trigerring any other jobs.
```
[the remark this answers is not in the document]
Comment (Didainius) [reply govcd/task.go:476]: fixed.

File: types/v56/ip_space.go [hunk 10]
```diff
@@ -200,3 +200,119 @@ type IPPrefixSequence struct {
 	// Specifies the percentage of allocated IP prefix blocks out of total specified IP prefix blocks.
 	AllocatedPrefixPercentage float32 `json:"allocatedPrefixPercentage,omitempty"`
 }
+
+// IpSpaceUplink sSpecifies the IP Space Uplink configuration for Provider Gateway (External network
```
[the remark this answers is not in the document]
Comment (Didainius) [reply types/v56/ip_space.go:204]: fixed

File: types/v56/ip_space.go (hunk 7 above)
Comment (Didainius) [reply types/v56/ip_space.go:222]: I removed this particular limitation, as it might age too quick

**Review by dataclouder — APPROVED** [review]

File: govcd/ip_space_allocation.go [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+
+	queryParams := queryParameterFilterAnd(fmt.Sprintf("type==%s", allocationType), queryParameters)
+	typeResponses := []*types.IpSpaceIpAllocation{{}}
+	err = client.OpenApiGetAllItems(apiVersion, urlRef, queryParams, &typeResponses, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Wrap all typeResponses into DefinedEntityType types with client
```
Comment (adambarreiro) [inline govcd/ip_space_allocation.go:80]:
```suggestion
	// Wrap all typeResponses into IpSpaceIpAllocation types with client
```

File: govcd/ip_space_allocation.go [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+
+	return nil
+}
+
+func allocateIpSpaceIp(client *Client, orgId, orgName, ipSpaceId string, ipAllocationConfig *types.IpSpaceIpAllocationRequest) ([]types.IpSpaceIpAllocationRequestResult, error) {
+	if orgId == "" || orgName == "" || ipSpaceId == "" {
+		return nil, fmt.Errorf("IP Space must have all values Org ID, Org Name and IP Space ID populated")
+	}
+
+	// client := ipSpace.vcdClient.Client
```
Comment (adambarreiro) [inline govcd/ip_space_allocation.go:205]: Leftover?

File: govcd/ip_space_allocation.go [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+		return nil, fmt.Errorf("error getting tenant context: %s", err)
+	}
+
+	typeResponses := []*types.IpSpaceIpAllocation{{}}
+	err = client.OpenApiGetAllItems(apiVersion, urlRef, queryParameters, &typeResponses, getTenantContextHeader(tenantContext))
+	if err != nil {
+		return nil, err
+	}
+
+	// Wrap all typeResponses into DefinedEntityType types with client
```
Comment (adambarreiro) [inline govcd/ip_space_allocation.go:270]:
```suggestion
	// Wrap all typeResponses into IpSpaceIpAllocation types with client
```

File: govcd/ip_space_org_assignment.go [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+		return nil, err
+	}
+	queryParams := queryParameterFilterAnd(fmt.Sprintf("ipSpaceRef.id==%s", ipSpace.IpSpace.ID), queryParameters)
+	typeResponses := []*types.IpSpaceOrgAssignment{{}}
+	err = client.OpenApiGetAllItems(apiVersion, urlRef, queryParams, &typeResponses, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Wrap all typeResponses into IpSpace types with client
```
Comment (adambarreiro) [inline govcd/ip_space_org_assignment.go:50]:
```suggestion
	// Wrap all typeResponses into IpSpaceOrgAssignment types with client
```

File: govcd/ip_space_uplink.go [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+		return nil, err
+	}
+
+	typeResponses := []*types.IpSpaceUplink{{}}
+	err = client.OpenApiGetAllItems(apiVersion, urlRef, queryparams, &typeResponses, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Wrap all typeResponses into DefinedEntityType types with client
```
Comment (adambarreiro) [inline govcd/ip_space_uplink.go:76]:
```suggestion
	// Wrap all typeResponses into IpSpaceUplink types with client
```

File: govcd/tenant_context.go [hunk 16]
```diff
@@ -188,6 +188,16 @@ func (vdcGroup *VdcGroup) getTenantContext() (*TenantContext, error) {
 	return org.tenantContext()
 }
 
+// getTenantContext retrieves the tenant context for a VdcGroup
```
Comment (adambarreiro) [inline govcd/tenant_context.go:191]:
```suggestion
// getTenantContext retrieves the tenant context for a IpSpaceIpAllocation
```

**Review by adambarreiro — APPROVED** [review]
Great PR, just some really minor things

File: govcd/tenant_context.go (hunk 16 above)
Comment (Didainius) [reply govcd/tenant_context.go:191]: Fixed.

File: govcd/ip_space_uplink.go (hunk 15 above)
Comment (Didainius) [reply govcd/ip_space_uplink.go:76]: Fixed

File: govcd/ip_space_org_assignment.go (hunk 14 above)
Comment (Didainius) [reply govcd/ip_space_org_assignment.go:50]: Fixed

File: govcd/ip_space_allocation.go (hunk 13 above)
Comment (Didainius) [reply govcd/ip_space_allocation.go:270]: Fixed

File: govcd/ip_space_allocation.go (hunk 12 above)
Comment (Didainius) [reply govcd/ip_space_allocation.go:205]: Fixed.

File: govcd/ip_space_allocation.go (hunk 11 above)
Comment (Didainius) [reply govcd/ip_space_allocation.go:80]: Fixed.

**Pull request merged**